### DISCUSSION TITLE
parabéns e dúvidas

### DISCUSSION THREAD

**Discussion by zepadovani** [charlesneimog/py4pd#58] [General]
A lib tá muito legal Charles!! So agora to tendo tempo de ver mais de perto! Muitas possibilidades abertas!!! Parabéns!

Vou sintetizar aqui, em alguns pontos, algumas dúvidas, tudo bem?

1.
Não consigo abrir o editor ao clicar na caixa, mesmo após setar. Estou no Manjaro, acho que isso é possivelmente relevante. Testei com diferentes comandos em 'editor <comando>' de binários do meu sistema. Quando clico no objeto ele até imprime "[py4pd] Opening editor...", mas não acontece nada.

2. Vi que há um import pd no py.py, onde fica esse pacote? Ele é embarcado de alguma maneira nas bibliotecas dinâmicas geradas? Queria entender melhor como você acessa métodos do pd... Talvez não esteja entendendo bem a arquitetura, mas é que realmente me interessei mais em cruzar as informações do python para o PD agora e queria entender melhor a estrutura do código. Enfim, talvez possa cooperar também com o projeto (já que estamos usando muito PD e Python aqui), mas nesse momento queria compreender melhor.

3. Outra dúvida, é possível especificar um environment venv ou conda? Pelo que vi ele está tentando acessar o Python geral. Para isso criei um script como esse:

```
import pd
import sys

def info():
    '''Prints information about the module.'''
    pd.print(f"Python version: {sys.version}")
    pd.print(f"Python binary: {sys.executable}")
    return 0
```

```
[Python]: Python version: 3.10.11 (main, May 16 2023, 00:37:51) [GCC 11.2.0]
[Python]: Python binary: /usr/bin/python3
```

No miniconda e com o venv, geralmente o binário fica em outro caminho... seria possível setar o ambiente?

No mais, parabéns novamente!! Realmente impressionado! Ferramenta muito promissora!
abraços!!

**Comment by zepadovani** [comment]
pequeno adendo, na última versão, na verdade, mesmo após setar o binário, o  py4pd parece não ter conseguido setar adequadamente o caminho... ele primeiro diz ter setado mas depois não consegue abrir.

```[py4pd] Editor set to: nvim
[py4pd] To open the editor you need to set the function first!
[py4pd] Editor set to: vscode
[py4pd] To open the editor you need to set the function first!
[py4pd] Editor set to: code
[py4pd] To open the editor you need to set the function first!
```

**Comment by charlesneimog** [comment]
Ei Zé, que bom que esta usando, se tiver dando tudo errado me avise, quero muito deixar bem estável para poder desenvolver minhas pesquisas em Python e poder compartilhar com pessoal que so usa Pd. Te respondendo.

1. Seguindo principalmente as mensagem do segundo comentário, você está setando o editor mas não a função que o objeto vai usar. Dessa forma o objeto não sabe que arquivo abrir. Se vc quer criar um novo arquivo você precisa usar a mensagem `open MeuArquivoPython` sem `.py`. Isso vai criar e abrir um arquivo onde seu patch está salvo. Veja se é isso, essa questão de editor e bem chata de configurar, se n for me avise pf. Última coisa, o nome para abrir o vscode é `vscode`. _Eu vi que eu tirei vários métodos help patch, eu vou colocar de novo._

2. Sim, o modulo `pd` é um modulo interno do objeto, ele é escrito em C e fica em `src/module.c`, penso ele principalmente para interagir com o pd e para criar objeto (não sei se já viu a função `pd.add_object` examplo de uso [orchidea](https://github.com/charlesneimog/orchidea/blob/6d1d351fb049e3b6722ef089e793a7aaa2aff7a6/orchidea.py#L95)). O modulo está com a documentação disponível aqui https://charlesneimog.github.io/py4pd/python-users/, se sentir falta de algo me avise também. 

3. Sim, é possível usar um ambiente conda, no meu caso eu faço isso usando a seguinte mensagem no objeto `packages /home/neimog/miniconda3/lib/python3.11/site-packages` (não é a melhor forma). Se você sempre quer usar o mesmo ambiente, você pode criar um arquivo chamado `py4pd.cfg` dentro da pasta do externo (MAIS CONFIÁVEL), dessa forma:

`py4pd.cfg` Vou colocar isso na documentação.
```
editor = nvim
packages = /home/neimog/miniconda3/envs/composition/lib/python3.11/site-packages/
```
Mas importante, a versão do python do ambiente precisa ser a mesma do `py4pd`, isso por que `pip` instala binários diferentes para cada versão, portanto, bibliotecas como librosa não vão funcionar caso a versão de Python seja diferente.   

Você também pode criar um ambiente local para o objeto (disponível para todo seus patches) e para os patches (só para o patch que está trabalhando). Para o objeto, você pode mandar a mensagem `pipinstall global NOMEDOMODULO` para `py4pd`, para os patches você pode usar `pipinstall local NOMEDOMODULE`, o módulo fica disponível para todos os patches na mesma pasta.

É isso, qualquer coisa me fale!!! O que mais preciso agora é do pessoal testando para eu saber e resolver os bugs!!! 

EDIT: Mais um coisa super importante, dentro do https://github.com/charlesneimog/py4pd/tree/master/resources/scripts, tem um módulo escrito em Python que chama py4pd. Nele tem objeto como `py.iterate`, `py.collect`, etc. Minha intenção é imitar o OpenMusic. Você pode usar esses objetos criando um objeto `py4pd -lib py4pd`. Se puder/quiser contribuir para essa biblioteca, estou tentanto reescrever os principais objetos de OM em Python.

Abrs

**Comment by zepadovani** [comment]
Obrigado Charles! O editor tá funcionando, agora entendi! inclui no cfg e está ótimo...
Já o environmente, parece que ainda assim não rola.

Meu cfg está assim

```
editor = vscode
packages = /home/padovani/miniconda3/envs/nospassarinho_3.11.4/lib/python3.11/site-packages/
```

Também tentei o comando que packages com o caminho acima, mas quando rodo minha função de teste, o python binary continua apontando para o binário do sistema (e não para o do meu ambiente). 

Para me certificar também tentei carregar pacotes que só tenho instalados no ambiente em questão e, de fato, dá erro.

Agluma dica?

Ah, vou testar os objetos sim!

**Reply by charlesneimog** [reply]
Acho que entendi, errei como eu gero esse `.cfg`, toda vez que vc usa `editor vscode` ele salva o editor nesse arquivo, mas eu esqueci de, se tiver alguma outra informação, reescrever também rsrs.

Logo, se vc criar esse arquivo como o que você mostrou, e NÃO usar a função `editor`  ou `packages`, ele vai funcionar. Ele vai ler que vc usa o vscode e o seus packages. Não precisa setar de novo.

Aqui para mim funcionou, me fale se ai também. Vou consertar isso e publicar uma nova versão.

Outra coisa, ele vai ler o executavel python principal. Ela não está rodando o python em /usr/bin/python3.11 mas lendo a biblioteca `python3.11.so`. Entretando, nessa biblioteca tem a informação que o executavel é o python do sistema, isso e fake rsrs, o que executa os códigos é o objeto, não outro executavel. A questão é que quando você configura os packages ele deve conseguir ver e ler todos os modulos instalados no conda. 

Qualquer coisa me fale.

**Comment by zepadovani** [comment]
Mas é isso que não ocorre... o vscode, sim, ele reconhece como esperado
(sem precisar usar mais o comando editor). Mas o Python continua
configurado naquele do sistema.... (independemente do que estiver no cfg)

Aí tentei usar o comando packages e tirar o cfg da pasta, mas dá o mesmo
erro... continua sempre usando o python do sistema.

**Reply by charlesneimog** [reply]
Mas vc também não consegue carregar os packages do conda?

 Por que quando vc usa `sys.executable`, ele pega essa informação de `python3.11.so`. Por estabilidade, o python que uso para carregar o objeto é essa `.so` do sistema mesmo, MAS, os packages, o versionamento e tudo mais e controlado pelo objeto `py4pd` não pelo executavel. 

A não ser que você queria fazer multithreading, isso nao deve dar problemas. Você vai conseguir usar o packages do seu conda normalmente.

**Reply by zepadovani** [reply]
pois é... mas os packages não estavam carregando não.  Só com a gambiarra abaixo, que é incluir o caminho para os pacotes do meu environment no meu script, usando o `sys.path.append`

por exemplo (rodando o abjad, que não tenho instalado no python do sistema mas tenho no meu environment):

```
import pd
import sys

# Add the path to your custom directory
custom_directory = '/home/padovani/miniconda3/envs/nospassarinho_3.11.4/lib/python3.11/site-packages/'
sys.path.append(custom_directory)

import abjad
import abjadext.nauert as nauert
import numpy as np
import os

def scorerun():
    rh_voice = abjad.Voice(name="RH_Voice")
    rh_staff = abjad.Staff([rh_voice], name="RH_Staff")
    lh_voice_1 = abjad.Voice(name="LH_Voice_1")
    lh_voice_2 = abjad.Voice(name="LH_Voice_2")
    lh_staff = abjad.Staff(name="LH_Staff", simultaneous=True)
    lh_staff.extend([lh_voice_1, lh_voice_2])
    piano_staff = abjad.StaffGroup(lilypond_type="PianoStaff", name="Piano_Staff")
    piano_staff.extend([rh_staff, lh_staff])
    score = abjad.Score([piano_staff], name="Score")
    rh = r"""
        a'8 \pp [ ( g' f' e' ] d'4 )
        g'8 \mp [ ( f' e' d' ] c'8 \< d'16 e' f'8 e' \! d'2 \> ~ d' \! )
    """
    lh_1 = r"s1 * 2/4 s1 * 3/4 s1 * 2/4 \voiceOne b2 ~ b"
    lh_2 = r"""
        b4 d'8 \pp [ ( c' b a ] af4 )
        c'8 \mp [ ( bf a g ] fs g16 a16 \voiceTwo b4 a g2 )
    """
    rh_voice.extend(rh)
    lh_voice_1.extend(lh_1)
    lh_voice_2.extend(lh_2)
    time_signature = abjad.TimeSignature((2, 4))
    note = abjad.select.note(rh_voice, 0)
    abjad.attach(time_signature, note)
    time_signature = abjad.TimeSignature((3, 4))
    note = abjad.select.note(rh_voice, 4)
    abjad.attach(time_signature, note)
    time_signature = abjad.TimeSignature((2, 4))
    note = abjad.select.note(rh_voice, 9)
    abjad.attach(time_signature, note)
    clef = abjad.Clef("bass")
    note = abjad.select.note(lh_voice_2, 0)
    abjad.attach(clef, note)
    bar_line = abjad.BarLine("|.")
    note = abjad.select.note(lh_voice_2, -1)
    abjad.attach(bar_line, note)
    note = abjad.select.note(rh_voice, -2)
    abjad.override(note).Hairpin.to_barline = False
    abjad.show(score)
    return 0
```

**Reply by charlesneimog** [reply]
Ah legal, agora consegui reproduzir. 

Mas so tenho o erro com a biblioteca `abjadext.nauert`, você também? Ou com desde abjad?

**Reply by zepadovani** [reply]
Na verdade no código que enviei nem tava usando ela... mas nesse que vou te mandar mais abaixo, por exemplo, tudo roda tranquilo. Mas a questão é que as bibliotecas (abjad-ext-nauert e abjad ) só estão rodando por conta das linhas
```
custom_directory = '/home/padovani/miniconda3/envs/nospassarinho_3.11.4/lib/python3.11/site-packages/'
sys.path.append(custom_directory)
```

que, na prática, permitem ao python encontrar os pacotes do meu ambiente. Pelo que entendi, usar o comando packages ou configurar o cfg com a variável packages deveria dar conta de realizar exatamente a mesma coisa, sem a necessidade disso... ou entendi mal?

Segue o código que funciona pra mim (usando abjad e o nauert):

```
import pd
import sys

# Add the path to your custom directory
custom_directory = '/home/padovani/miniconda3/envs/nospassarinho_3.11.4/lib/python3.11/site-packages/'
sys.path.append(custom_directory)

import abjad
import abjadext.nauert as nauert
import numpy as np
import os

def scorerun():
    durations = np.random.randint(700, 1200, 8)
    pitches = range(8)
    q_event_sequence = nauert.QEventSequence.from_millisecond_pitch_pairs(
        tuple(zip(durations, pitches))
    )
    result = nauert.quantize(q_event_sequence)
    staff = abjad.Staff([result])
    score = abjad.Score([staff])
    abjad.show(score)  
```

**Reply by charlesneimog** [reply]
É, é isso mesmo. Deveria dar conta. 

Aqui eu retirei as linhas que vc coloca o ambiente so sys.path. A abjad carrega, mas a outra não. Se puder testar se acontece o mesmo para você, eu agradeço.

Sobre a `abjadext.nauert` eu vou ver o que é. Eu acredito que como to salvando esses dados em py4pd.cfg talvez nao seja a melhor forma. 

Vc tem acesso os objetos compilados em `Actions`? Se sim, quando tiver algo que ache que resolva te marco aqui.

**Reply by zepadovani** [reply]
Se eu tiro esse comando do sys.path, nenhum pacote do meu ambiente é carregado (o abjad já dá erro).

Esse 'nauert' tem um pacote próprio, é um quantizador para o abjad... (pip install abjad-ext-nauert)

Sobre o Actions, acho que só aquilo que você manda ir pra release (https://github.com/charlesneimog/py4pd/tags) não sei se foi assim que vc configurou... ?

**Reply by charlesneimog** [reply]
Não, esses daqui:

https://github.com/charlesneimog/py4pd/actions/runs/6961694965 (se for igual ao meu, tem que ter uma py4pd que da para baixar).

![image](https://github.com/charlesneimog/py4pd/assets/31707161/ab9ac096-a875-4db7-a9fd-3ff012fe4971)

Esse é o objeto do branch dev. 

Eu vou tentar resolver isso, mas acho que é relacionado com o py4pd.cfg, pq quando eu mando `pipinstall local abjad-ext-nauert` para o objeto `py4pd` eu consigo carregar e rodar a sua função. Ele só não vai conseguir mostrar com o show.

Eu te respondo assim que conseguir...


Mas muito obrigado pelo feedback!

**Reply by zepadovani** [reply]
eu que agradeço!! Tá ficando muito bacana essa biblioteca... expande demais as possibilidades do PD! 
Qq coisa a gente marca um zoom depois? Porque são várias questões aqui que foram surgindo, dúvidas e tal...

abração e muito obrigado pela ajuda!

**Reply by zepadovani** [reply]
ah, tenho acesso sim!

**Reply by charlesneimog** [reply]
Sim, podemos marcar um zoom... Pra mim é ótimo que vou afinando o objeto e entendo como isso funciona nas máquinas de outras pessoas. 

Tendo resolver isso até semana que vem...

**Reply by charlesneimog** [reply]
Meu e-mail é charlesneimog@outlook.com, se quiser mandar algo por la!

**Comment by charlesneimog** [comment]
Oi Bom dia!

Vc pode ver se essa build resolve o problema para vc.

https://github.com/charlesneimog/py4pd/actions/runs/6970272238

🤞️🤞️🤞️🤞️🤞️

Se não resolver, vc pode fazer me passar duas informações?

1. Fazer uma função com pd.print(sys.path) do pd.
2. Me passar as informações de sys.path do seu ambiente.

Veja tb se no seu sys.path, se não tem nada relacionado com a abjad.

Outra coisa, o `abjad.show(score)` não vai funcionar, ele da crash no Pd. Eu acho que é por causa do Tk, mas esse é um problema de conflito, ainda n sei como resolver ele. 

Para uma outra solução, vc pode retornar um path com uma imagem .png e mostrar ela usando o objeto `py.show`. Que fica disponivel depois que vc criar um outro objeto chamado `py4pd -lib py4pd`.

Ficaria tipo isso.

https://github.com/charlesneimog/py4pd/blob/eebf0325512d8a0113520c5ab269cd971f2294ca/docs/examples/score/score.gif

**Reply by zepadovani** [reply]
Bom dia!!

Deu certo! Super obrigado!! 
Ah, e o abjad.show(score)  funciona aqui sim, sem problemas!  (talvez vc tenha que configurar ele aí, que é um caso a parte)

Mas muito legal, de qualquer maneira, esse objeto py.show!!

**Reply by charlesneimog** [reply]
Amém igreja!!! rsrsrs
